pygin/pygin/src/helpers.cpp

```diff
@@ -0,0 +1,52 @@
+#include "headers.hpp"
+
+#include "helpers.hpp"
+#include "python.hpp"
+#include "py.common.hpp"
+
+namespace helpers
+{
+
+namespace {
+struct py_deleter
+{
+	void operator ()(void* p) const { PyMem_Free(p); }
+};
+} // anonymous
+
+void py_string_storage::clear()
+{
+	while (!this->strings.empty())
+	{
+		py_deleter()(this->strings.back());
+		this->strings.pop_back();
+	}
+}
+
+wchar_t const* py_string_storage::push_back(py::string const& s)
+{
+	Py_ssize_t Size;
+	std::unique_ptr<wchar_t [], py_deleter> Data(
```

Comment:
This is very cryptic.
Why do we bother with a smart pointer and a custom deleter if we later throw it away, store the raw pointer and later call the deleter manually?
Why don't have a vector of smart pointers, or even just strings?

Supposedly the concern is pointer invalidation on subsequent pushes, due to reallocations? If so, the underlying container could be changed to list. [Example](https://github.com/FarGroup/FarManager/blob/6edbccc87c51a4d9900bdcc7586f95a9395c88b9/far/plugsettings.cpp#L231-L256). Anything is better than manual memory management.

Reply:
- `std::unique_ptr` is used for exception safety - `strings.push_back` might throw.
- `std::vector<wchar_t*>` is needed to be able to return `wchar_t const* const*` to Far API. See https://github.com/trexinc/evil-programmers/pull/51/files#diff-d81348ec57cdb59f76d82acb398fcba41bf476a6612952143dd2441692c1d797R252